src/kubernetes_operator/iam_mapping.py

```diff
@@ -0,0 +1,223 @@
+"""Kubernetes operator to manage IamIdentityMappings in the aws-config configmap."""
+import asyncio
+import logging
+from pathlib import Path
+
+import kopf
+import yaml
+from kubernetes import client, config
+from kubernetes.client.models.v1_config_map import V1ConfigMap
+
+logger = logging.getLogger("operator")
+
+try:
+    config.load_kube_config()
+except Exception as error:
+    logger.info("Could not load kubeconfig. Error is : %s.\nAssuming we are in a kubernetes cluster", error)
+    try:
+        config.load_incluster_config()
+    except Exception as error:
+        error_str = format(error)
+        raise Exception(f"No k8s config suitable, exiting ({error_str})") from error
+else:
+    logging.info("Using Kubernetes local configuration")
+
+API = client.CoreV1Api()
+custom_objects_API = client.CustomObjectsApi()
+GROUP = "iamauthenticator.k8s.aws"
+VERSION = "v1alpha1"
+PLURAL = "iamidentitymappings"
+
+
+@kopf.on.update(GROUP, VERSION, PLURAL)
+@kopf.on.create(GROUP, VERSION, PLURAL)
+async def create_mapping(spec: dict, diff: list, **_) -> None:
+    """Create/update an identity mapping in the aws-auth configmap with the corresponding IamIdentityMapping.
+
+    This method accepts mappings for userarn and rolearn with groups.
+
+    :param spec: The spec of the changed IamIdentityMapping
+    :param diff: The diff created by the changed identity
+    """
+
+    # Do nothing when we have no diff
+    if not diff:
+        return
+
+    sanitize_spec = dict(spec)
+    configmap = API.read_namespaced_config_map("aws-auth", "kube-system")
+
+    arn_field = spec["userarn"] if spec.get("userarn") else spec["rolearn"]
+    logger.info("Mapping for user %s as %s to %s", arn_field, spec["username"], spec["groups"])
+
+    identities = get_identity_mappings(configmap)
+    updated_mapping = ensure_identity(sanitize_spec, identities)
+    await apply_identity_mappings(configmap, updated_mapping)
+
+
+@kopf.on.delete(GROUP, VERSION, PLURAL)
+async def delete_mapping(spec: dict, **_) -> None:
+    """Delete the identity mapping in the aws-auth configmap corresponding to the deleted IamIdentityMapping.
+
+    :param spec: The spec of the removed IamIdentityMapping
+    """
+    configmap = API.read_namespaced_config_map("aws-auth", "kube-system")
+
+    arn_field = spec["userarn"] if spec.get("userarn") else spec["rolearn"]
+    logger.info("Delete mapping for user %s as %s to %s", arn_field, spec["username"], spec["groups"])
+
+    identity_mappings = get_identity_mappings(configmap)
+    updated_mappings = delete_identity(spec, identity_mappings)
+    await apply_identity_mappings(configmap, updated_mappings)
+
+
+@kopf.on.startup()
+def on_startup(logger, **_) -> None:
```

Comment:
I just did a small test:
- Apply a aws-auth CM in the kube-system Namespace with stuff in it (stuff you won't find in a IAMIdentityMapping), like ours here https://github.com/coveo/infra-kubernetes/blob/7cde0cc290558dbd36eac3ff22d2a87a23466010/infrastructure/terraform/coveo-eks/aws_auth_cm.tf
- Boot this Operator
- Watch the operator remove everything

Booting this Operator in a live Coveo EKS cluster will effectively brick the cluster 💥 

We need to take into account this aws-auth CM will not be empty, or we need to create one IAMIdentityMapping in the infra-kubernetes repo for each entry in our detault CM.

I'll let your decide with the Infra team.

Reply:
To explain further, the original operator overrides mapUsers with what it finds in the IamIdentityMappings, even if there were other entries unrelated to IamIdentityMappings (which didn't matter since there is nothing else there currently).
 
Transferring this logic to mapRoles would then also make it override everything including role mappings that are there and unrelated to the IamIdentityMappings.

Reply:
These roles are there for a reason, we need access while UAM didn't pass yet. Why would we want to remove them?

Reply:
Hi Dave,

These roles might be useful when UAM did not pass yet, but UAM is passing quite quickly and in my opinion, it is not an issue.

In addition to by-passing our user access management system, the way the roles mapping is done at the moment is making it impossible to retrace who did what in the cluster, because in the Kubernetes audit trail, we won't see "Dave did X", we see "kubectl-access-user-infra did x".

With strongDM, the audit trail is not an issue, because strongDM becomes the owner of that audit trail.

To recap, having a way to bypass our access management system that also breaks the required traceability is not ideal in my opinion.

I would:
- Add code to ignore the `system:node:{{EC2PrivateDNSName}}` user of the `mapRoles` list
- Flush all other entries that are not defined in a proper `IAMIdentityMapping` object

That being said, if you absolutely want to keep your UAM bypass, simply setup a `IAMIdentityMapping` outside of the `aws-auth` ConfigMap and our Operator will apply it.

So, what should we do? 😄

Reply:
"That being said, if you absolutely want to keep your UAM bypass, simply setup a IAMIdentityMapping outside of the aws-auth ConfigMap and our Operator will apply it." 

This makes it possible for test cluster, etc. so ia have nothing to add. Its all good for me.

As for the actual "breaking" situation, i suppose it is feasable to do a progressive rollout that tolerate both for a while?

Reply:
Tristen can do a PR in infra-kubernetes to add the IAMIdentityMapping you require.

On the first UAM pass, UAM will delete the IAMIdentityMapping to make the cluster compliant.

Then we can remove hardcoded access rights in the aws-auth CM, upgrade the Operator.

Good?

Reply:
est-ce que ca pourrait affecter le deploy_role que terraform utilise pour les provider k8s et helm?

Reply:
La mécanique entre Terraform et Kubernetes m'est inconnu.

Est-ce que @jocgir peux nous en dire plus?

Reply:
En faite, la coupure sera au moment de la migration, donc fort probablement négligeable. votre plan semble etre solide :+1:

Reply:
To make it work, there a lot of details to take into consideration. I am currently working on a plan to make sure everything runs smoothly. I'll post back next week when I've tested everything I could and I came up with a solid migration plan.